### PR TITLE
Retry requests for reliability on 429 status [ch9191]

### PR DESCRIPTION
**Pull request by bogdanguranda**
Ticket: [[ch9191]](https://app.clubhouse.io/chartmogul/story/9191/retry-on-429-from-cb)

### CODE REVIEW

File: http_request.go [hunk 1]
```diff
@@ -15,10 +17,35 @@ func Do(req *http.Request, env Environment) (string, error) {
...
+	var response *http.Response
+	var err error
+	// Retry on HTTP 429 rate limit, or network error, see:
+	// https://godoc.org/github.com/cenkalti/backoff#pkg-constants
+	backoff.Retry(func() error {
+		response, err = client.Do(req)
+		if err != nil {
+			return nil
+		}
+		if response == nil || response.StatusCode == http.StatusTooManyRequests {
```
Comment (hassansin) [inline http_request.go:40]: Are you implying network error from `response== nil`?

**Review by hassansin — APPROVED** [review]

File: http_request.go (hunk 1 above)
Comment (pkopac) [reply http_request.go:40]: You are right, we should instead retry if there _is_ `err`.
> A non-2xx response doesn’t cause an error.

But, keep it saved to return after last retry.

We took it from here, but not perfectly, https://github.com/chartmogul/chartmogul-go/blob/be00f9b814ec5d1ab64fe2e4ff88e3e8df850718/generic.go#L29

Comment (hassansin) [reply http_request.go:40]: You can borrow from here: https://github.com/chartmogul/go-utilities/blob/a9ddca6f22358acdc1fbeeb38837ac578de81559/utils/errors.go#L112

**Review by pkopac — APPROVED** [review]
@hassansin we used the != url.Error

**Pull request merged**